Implement the estimation of maximum number of inputs for Shelley and show big wallet migration

paweljakubas opened the pull request:
# Issue Number

<!-- Put here a reference to the issue this PR relates to and which requirements it tackles -->
#1751 

# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- [x] I have refactored computeTxSize to use later
- [x] I have implemented `_estimateMaximumNumberOfInputs`
- [x] I have unpended integration test showing the case - big wallet migration
- [x] I have added property tests

# Comments

<!-- Additional comments or screenshots to attach if any -->

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Assign the PR to a corresponding milestone
 ✓ Acknowledge any changes required to the Wiki
-->

paweljakubas committed:
make computeTxSize available outside minimumFee

paweljakubas committed:
implement estimateMaxNumberOfInputs

paweljakubas committed:
correcting estimation and un-pending integration test

paweljakubas committed:
add property and make some performance optimisation

hlint

piotr-iohk commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Migrations.hs:
```diff
@@ -246,10 +242,9 @@ spec = do
             [ expectField
                 #distribution
-                ((`shouldBe` (Just 400)) . Map.lookup 10_000_000_000)
+                ((`shouldBe` (Just 129)) . Map.lookup 100_000_000_000)
```
This does not appear to be right... :thinking: 
The fixture wallet has 400 utxo * 100_000_000_000 + dust (100 utxo * 1), so I'd expect target wallet ditribution to be 400, not 129...

https://github.com/input-output-hk/cardano-wallet/blob/master/lib/shelley/test/data/cardano-node-shelley/genesis.yaml#L1217

paweljakubas replied:
I do not confirm original balance 400 * ...
I see : 
```
################## originalBalance:12900000000000
################## expectedFee:3574500
################## expectedBalance:12899996425500
```

piotr-iohk replied:
Yes, but wallet from the genesis file has the following: 400 * 100_000_000_000 and 100 * 1... so  the original balance should be 40000000000500... :thinking:

paweljakubas replied:
I think it should be subject of another BUG ...

KtorZ replied:
The problem comes from:

a) Funds are defined as a key-value "map" in the genesis file. 
b) This is therefore parsed as a map by ledger's functions & al. 
c) These identify _sequential_ addresses, that need to be discovered in order! However, the map "destroys" any sort of order we could have on keys... and there's no guarantee that we process addresses in order. Actually, we don't, and are only able to discover 129 of those addresses.

KtorZ replied:
Was harder than I thought, but I got it:

```
(Status {statusCode = 200, statusMessage = "OK"},Right (ApiUtxoStatistics {total = Quantity {getQuantity = 40000000000100}, scale = ApiT {getApiT = Log10}, distribution = fromList [(10,100),(100,0),(1000,0),(10000,0),(100000,0),(1000000,0),(10000000,0),(100000000,0),(1000000000,0),(10000000000,0),(100000000000,400),(1000000000000,0),(10000000000000,0),(100000000000000,0),(1000000000000000,0),(10000000000000000,0),(45000000000000000,0)]}))
```

paweljakubas committed:
rebase conflict fix

KtorZ committed:
rewrite function and variables in estimateTxMaxSize to improve readability

Functions were quite hard to follow (and weirdly indented). Took me a while to figure out what the algorithm was about while reviewing.
So I rename a few things without changing the logic to make it more obvious at first glance.

KtorZ committed:
do not use implementation internal functions in property test for tx max size

The property was written by using the same function used for implementing the function it was testing. This is a recipe for errors for
the property will have the same bugs as the underlying function. Instead, I've rewritten properties to use only the outter function and
make statement about what can be observed. Also added some hard-coded classic unit tests to control the order of magnitude for the number
of inputs.

KtorZ committed:
preserve order of initial funds when parsing genesis file

KtorZ committed:
use correct value in integration scenario with big wallet + increase request timeout to 180s

I don't think increasing the timeout is _the right_ answer here, I'd
rather fix the endpoint so that it could return an answer to client
immediately while submitting transactions in the background.

Commit:
Regenerate nix

KtorZ committed:
make the 'big' wallet a bit less ... big.
